pkg/eventsources/sources/azureeventshub/start.go

```diff
@@ -61,87 +63,106 @@ func (el *EventListener) GetEventSourceType() aev1.EventSourceType {
 			return err
 		}
-		return nil
-	}
-
-	// listen to each partition of the Event Hub
-	log.Info("gathering the hub runtime information...")
-	runtimeInfo, err := hub.GetRuntimeInformation(ctx)
-	if err != nil {
-		return fmt.Errorf("failed to get the hub runtime information for %s, %w", el.GetEventName(), err)
-	}
 
-	if runtimeInfo == nil {
-		return fmt.Errorf("runtime information is not available for %s, %w", el.GetEventName(), err)
+		log.Info("connecting to event hub using AAD credentials...")
+		consumerClient, err = eventhub.NewConsumerClient(hubEventSource.FQDN, hubEventSource.HubName, eventhub.DefaultConsumerGroup, credentials, nil)
+		if err != nil {
+			return fmt.Errorf("failed to connect to the hub %s, %w", hubEventSource.HubName, err)
+		}
 	}
 
-	if runtimeInfo.PartitionIDs == nil {
-		return fmt.Errorf("no partition ids are available for %s, %w", el.GetEventName(), err)
+	log.Info("retrieving event hub partitions...")
+	hubProps, err := consumerClient.GetEventHubProperties(context.Background(), nil)
+	if err != nil {
+		return fmt.Errorf("failed to get event hub properties: %w", err)
 	}
 
-	log.Info("handling the partitions...")
-	for _, partitionID := range runtimeInfo.PartitionIDs {
-		if _, err := hub.Receive(ctx, partitionID, handler, eventhub.ReceiveWithLatestOffset()); err != nil {
-			return fmt.Errorf("failed to receive events from partition %s, %w", partitionID, err)
-		}
+	for _, partitionID := range hubProps.PartitionIDs {
+		go func(partitionID string) {
+			log.Infow("connecting to partition", "partitionID", partitionID)
+			partitionClient, err := consumerClient.NewPartitionClient(partitionID, nil)
+			if err != nil {
+				log.Errorw("failed to create receiver for partition", "partitionID", partitionID, zap.Error(err))
+				return
+			}
+			defer partitionClient.Close(context.TODO())
+
+			for {
+				receiveCtx, cancel := context.WithTimeout(context.TODO(), 5*time.Second)
+				receivedEvents, err := partitionClient.ReceiveEvents(receiveCtx, 100, nil)
+				cancel()
+
+				if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+					log.Errorw("failed to receive events from partition", "partitionID", partitionID, zap.Error(err))
+					return
+				}
+
+				for _, event := range receivedEvents {
+					log.Infow("received event from partition", "eventID", *event.MessageID, "partitionID", partitionID)
+					eventData := &events.AzureEventsHubEventData{
+						Id:       *event.MessageID,
+						Body:     event.Body,
+						Metadata: hubEventSource.Metadata,
+					}
+
+					if event.PartitionKey != nil {
+						eventData.PartitionKey = *event.PartitionKey
+					}
+
+					eventBytes, err := json.Marshal(eventData)
+					if err != nil {
+						el.Metrics.EventProcessingFailed(el.GetEventSourceName(), el.GetEventName())
+						log.Errorw("failed to marshal the event data", "eventSource", el.GetEventName(), "messageID", *event.MessageID, zap.Error(err))
+						return
```

Comment:
This `return` will exit the goroutine for the corresponding partition, is it expected?

Reply:
Ah, good call! I suppose I can use `continue` instead of `return` here so that the goroutine does not exit abruptly. I can change that.